src/modules/mavlink/mavlink_parameters.cpp

```diff
@@ -560,6 +584,57 @@ MavlinkParametersManager::send_param(param_t param, int component_id)
 	return 0;
 }
 
+
+int MavlinkParametersManager:: send_error(MAV_PARAM_ERROR error, const char *param_id, const int param_index, const int target_sysid, const int target_compid,
+		int component_id)
+{
+       PX4_ERR("Debug:Test: %i", 1);
+	/* no free TX buf to send this param error message */
+	if (_mavlink.get_free_tx_buf() < MAVLINK_MSG_ID_PARAM_ERROR_LEN) {
+		return 1;
+	}
+        PX4_ERR("Debug:Test: %i", 2);
+	mavlink_param_error_t msg;
+	msg.target_system = target_sysid;
+	msg.target_component = target_compid;
+	msg.error = error;
+
+	if (param_index > -1) { // param_id is not used
+
+		msg.param_index = param_index;
+
+	} else {
+#if defined(__GNUC__) && __GNUC__ >= 8
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-truncation"
+#endif
+		/*
+		 * coverity[buffer_size_warning : FALSE]
+		 *
+		 * The MAVLink spec does not require the string to be NUL-terminated if it
+		 * has length 16. In this case the receiving end needs to terminate it
+		 * when copying it.
+		 */
+		strncpy(msg.param_id, param_id, MAVLINK_MSG_PARAM_ERROR_FIELD_PARAM_ID_LEN);
+#if defined(__GNUC__) && __GNUC__ >= 8
+#pragma GCC diagnostic pop
+#endif/* code */
+	}
+
+	/* default component ID */
+	if (component_id < 0) {
+		mavlink_msg_param_error_send_struct(_mavlink.get_channel(), &msg);
+
+	} else {
+		// Re-pack the message with a different component ID
+		mavlink_message_t mavlink_packet;
+		mavlink_msg_param_error_encode_chan(mavlink_system.sysid, component_id, _mavlink.get_channel(), &mavlink_packet, &msg);
+		_mavlink_resend_uart(_mavlink.get_channel(), &mavlink_packet);
+	}
```

Comment:
@julianoes This is supposed to send the error message. I copied it from the send_param() structure.

Thing is, I'm not seeing the emitted PARAM_ERROR when I send an invalid param but I am seeing the PX4 error log and evidence the code path is executed. Also, when I set a valid param I am not seeing the errors or the PARAM_VALUE message.

SO my inferences is that for whatever reason params are not being emitted on simulated channel. 
Am I doing something dumb here?

Reply:
Not sure. I'd have to test it. You also need to update mavlink somewhere in the pull request, I assume, or is it already updated?

Reply:
@julianoes FYI, this does work. Timing issue in my test. I'll clean it up and do more extensive tests.